### ISSUE TITLE
Code formatting causes cursor to jump to EOF

### ISSUE THREAD

**Issue by Cirzen** [PowerShell/vscode-powershell#1507]
I have my formatting style set to Allman. If I start typing something, e.g. a function, or a process block and put the open-brace at the end of the line, when the formatter kicks in and moves it to the next line, the cursor also jumps to the last line of the file.
Here's a screen grab of what happens.
![formatter](https://user-images.githubusercontent.com/13346199/44920339-94706300-ad37-11e8-9b66-d20c95c4247a.gif)

Disabling the add-in resolves the issue

### System Details

- Operating system name and version: Win 10 1709
- VS Code version: 1.26.1
493869ee8e8a846b0855873886fc79d480d342de
x64
- PowerShell extension version: 1.8.3
- Output from `$PSVersionTable`:
Name                           Value
----                           -----
PSVersion                      6.1.0-preview.1
PSEdition                      Core
GitCommitId                    v6.1.0-preview.1
OS                             Microsoft Windows 10.0.16299
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

### Issue Description

I am experiencing a problem with...

### Attached Logs
[EditorServices.zip](https://github.com/PowerShell/vscode-powershell/files/2340597/EditorServices.zip)
Follow the instructions in the [README](https://github.com/PowerShell/vscode-powershell#reporting-problems

)
about capturing and sending logs.

**Comment by Cirzen** [comment]
Cannot reproduce after a reboot. Closing as anomaly.

**Comment by tylerl0706** [comment]
Thanks for the follow up!

**Comment by ThubLives** [comment]
I see this behaviour pretty consistently with larger scripts and using OTBS formatting style.  Should I create a new issue with my info?

**Comment by rjmholt** [comment]
@ThubLives I think this issue is the best place. I will reopen.

Can you please follow the steps in our [troubleshooting docs](https://github.com/PowerShell/vscode-powershell/blob/master/docs/troubleshooting.md) and post the info/logs/any extra info you think might be relevant here?

**Comment by ThubLives** [comment]
Another quirk during cursor teleportation is that all folded code is left unfolded.  Also, sometimes when I find my way back to where my cursor should be, I notice that the formatting was incomplete.  For example, one remaining curly brace was not indented properly, though all the other lines were adjusted.

**Comment by andrewducker** [comment]
Yup.  This happens to me intermittently.  Usually when I hit "save" and it does an automatic format.

**Comment by ThubLives** [comment]
I now suspect that this is only happening in scripts that contain custom class definitions.  Can anyone else who's experienced this problem confirm this with their own scripts?  It may simply be that my larger scripts tend to use classes, but so far I haven't found any of my scripts without classes that trigger this behaviour.

**Comment by rjmholt** [comment]
If that's true, that's really great information!

**Comment by ThubLives** [comment]
Nope.  My theory seems to be a bust.  I found a 1690 line script that has no class definitions, but triggers this behaviour nonetheless.

**Comment by ThubLives** [comment]
With the release of vscode-powershell 1.10.0, the code-folding code has been removed from the client-side to the server-side.  Also with this version, whether related or not, I can no longer reproduce the cursor teleportation behaviour during automatic formatting, which has been a boon for my fraying sanity.  Even the unfolding of all code doesn't seem to be happening anymore, though this problem has been replaced with more subtle code folding bugs, deserving their own issue(s).

tl;dr:  No more unexpected cursor teleportation in v1.10.0 for me.  Code folding misbehaviour is different, but deserves a new issue.  Should this issue be closed?

**Comment by TylerLeonhardt** [comment]
Cool! Can anyone else repro this? @Cirzen @gerane @andrewducker

**Comment by rjmholt** [comment]
@ThubLives We're working on a fix for the folding issues here: https://github.com/PowerShell/PowerShellEditorServices/pull/805

**Comment by andrewducker** [comment]
I've not seen the issue today.

(I had been thinking that it was a clash with GitLens, which I started recently, but I turned that back on, and I'm still not seeing the issue.)

**Comment by andrewducker** [comment]
Aha!  That duplicated code issue is one I've seen repeatedly!  And once it happens, if I delete it then hitting "Save" (and triggering the automatic formatting) makes the code reappear.  Sometimes at the end of the file, sometimes randomly in the middle (possibly where the cursor was).

**Comment by andrewducker** [comment]
So presumably it only happens in files large (and/or complex) enough for the reformatting to take longer than it takes to update the file again?

**Comment by ThubLives** [comment]
@andrewducker I'm thinking you're right, this seems like some kind of race condition.  I have been largely free of these issues since updating to v1.10.0 (code folding has its own bugs since it's been moved to PS editor service), but I thought I would try GitLens on my current 2200 line script project.  I was almost immediately greeted with a familiar disorientation when Format on Type was triggered.

The cursor did not move and I didn't see any duplicate code, but all code was unfolded, which meant the text under the current view port moved and I was left looking at a much earlier portion of the script.  Fortunately, since the cursor was still at the same line number and column as before, I could just move the cursor with the arrow keys and reset the view port to continue working on the code.

I'm guessing that during formatting, the code is effectively unfolded until certain operations are completed, then re-folded at the user's selected points before the view port is updated.  It would seem that by installing GitLens, I have increased the duration of some operation triggered by formatting that happens between unfolding and re-folding code and is causing a race condition with some other operation that is interrupting the re-folding and possibly other operations.

In my case, only code folding is affected, but if the above is true, it would stand to reason that the cursor re-positioning might also be affected if the cursor is temporarily moved and is supposed to return to the appropriate line before the view port is updated, but is similarly interrupted.

As for the duplicated code, I'm having a hard time fitting that into my little theory.  I haven't seen that in my testing, so I'm not sure if there is any correlation to cursor position — other than where the duplicate code appears.

**Comment by lokwaniyash** [comment]
My html file is doing the same thing, It happens whether there are many lines or not, my 3000 line and 30 line files act the same. Any fixes cuz it's very annoying to click the cursor where the I'm typing after every save. Thanks.

**Comment by andrewducker** [comment]
If you've got a reproducible situation where it happens every time then that sounds incredibly useful!

**Comment by lokwaniyash** [comment]
As a matter of fact, it's happening right now as I'm playing with some stuff.

**Comment by andrewducker** [comment]
Does it still do it if you disable the PowerShell extension? Because if it's HTML then I'd expect it's not the extension doing it. Worth a test!

**Comment by lokwaniyash** [comment]
I don't even have it installed.

**Comment by andrewducker** [comment]
In that case, presumably not a plugin issue, but a VS Code issue.

**Comment by AshuGuptaGamer** [comment]
Have the same issue on one of my JS files

**Comment by andschwa** [comment]
This issue was for a bug in the PowerShell extension, solved a good while ago. Issues with HTML and JS files need to be filed with their respective extensions, thank you!
